### PR TITLE
Add initial work on composite-checkout package

### PR DESCRIPTION
**Pull request by sirbrillig**
This will be a new package that can be used to replace the checkout process in Calypso, CrowdSignal, Jetpack, and Akismet.

This is not used anywhere yet, but by merging into master it will become easier to iterate on the work.

### Screenshot

Screen Shot 2019-10-17 at 8 17 37 PM

### Testing instructions

To run the test server:

`npm install && npm run composite-checkout-demo`

Then visit http://localhost:3000/

### CODE REVIEW

**Comment by sirbrillig** [comment]
I'd kind of like to get some styles in here before we merge anything.

File: packages/wp-checkout/src/components/checkout-step.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+export default function CheckoutStep( {
+	className,
+	stepNumber,
+	title,
+	onEdit,
+	collapsed,
+	children,
+} ) {
+	const classNames = collapsed
+		? [ className, 'checkout-step', 'checkout-step--is-collapsed' ]
```
Comment (nbloomf) [inline packages/wp-checkout/src/components/checkout-step.js]: I think the host page could pass a space-delimited list for `className`. This is probably a good thing, maybe worth noting in the documentation.

Also- since we're using proptypes is it worth validating that `className` is a valid css class name?

File: packages/wp-checkout/src/components/checkout-step.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+				onEdit={ onEdit }
+			/>
+			{ children }
+		</div>
+	);
+}
+
+CheckoutStep.propTypes = {
+	className: PropTypes.string,
+	stepNumber: PropTypes.number.isRequired,
```
Comment (nbloomf) [inline packages/wp-checkout/src/components/checkout-step.js]: Should we validate which step numbers make sense? E.g. what if `stepNumber` is -1, or 1/2, or NaN?
If this is never exposed to the host page maybe it doesn't matter.

**Review by nbloomf — APPROVED** [review]
This is a great start.

This package is not live, but merging now and working off of master will make iteration much easier.

File: packages/wp-checkout/src/components/checkout-step.js (hunk 1 above)
Comment (sirbrillig) [reply packages/wp-checkout/src/components/checkout-step.js]: > I think the host page could pass a space-delimited list for className. This is probably a good thing, maybe worth noting in the documentation.

Good idea! Added to the README.

> since we're using proptypes is it worth validating that className is a valid css class name?

Hm... I suppose we could create a custom validator to do that. I'm not sure it's worth the effort, though. I think that if you're using `className` you probably know what kind of strings it supports. Also, CSS classNames can support a lot of kinds of strings.

File: packages/wp-checkout/src/components/checkout-step.js (hunk 2 above)
Comment (sirbrillig) [reply packages/wp-checkout/src/components/checkout-step.js]: > Should we validate which step numbers make sense? E.g. what if stepNumber is -1, or 1/2, or NaN?

That's a good idea! But I think I'd rather put that validation into the actual code, since propTypes only show a warning, and even then only in development. We can throw an Error if the prop is invalid.

> If this is never exposed to the host page maybe it doesn't matter.

It probably won't be when using the `Checkout` component, but `CheckoutStep` is going to be part of the public API so we should make it as helpful as possible.

**Comment by nbloomf** [comment]
> First of all, is the package name ok?

The hardest question. :) The `wp-` prefix to me seems to indicate something specific to WordPress or maybe WPCOM, which I suppose this sort of is in a de facto sense. The best alternative I can think of is `a8c-`.

As for `-checkout` -- I think that term is fine. My first thought is that something like `-pos` or `-point-of-sale` might make it more explicit that the package doesn't deal with cart management (whatever that means) but `-checkout` is not really confusing on this point either, and POS systems typically do stuff like inventory and relationship management that this package cares nothing about.

All that is to say, I can't think of a _better_ name for the package.

File: npm-shrinkwrap.json [hunk 3]
```diff
@@ -1140,7 +1367,6 @@
 			"version": "7.6.3",
 			"resolved": "https://registry.npmjs.org/@babel/traverse/-/traverse-7.6.3.tgz",
 			"integrity": "sha512-unn7P4LGsijIxaAJo/wpoU11zN+2IaClkQAxcJWBNCMS6cmVh802IyLHNkAjQ0iYnRS3nnxk5O3fuXW28IMxTw==",
-			"dev": true,
```
Comment (sirbrillig) [inline npm-shrinkwrap.json]: I'm not sure why these changes got made. All I've done is `npm install` after adding the new packages. Hopefully this is all ok.

File: packages/wp-checkout/src/components/checkout.js [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+	orderReviewTOS: PropTypes.element,
+	orderReviewFeatures: PropTypes.element,
+};
+
+function PaymentMethodsStep( { setStepNumber, isActive, isComplete, availablePaymentMethods } ) {
+	const localize = useLocalize();
+	const paymentMethod = usePaymentMethod();
+	const [ , setPaymentMethod ] = usePaymentMethodId();
+
+	// We must always display both the active and inactive version to keep their
```
Comment (fditrapani) [inline packages/wp-checkout/src/components/checkout.js]: Let's take a look at this after it merges. I don't like that we're duplicating the Step header along with the content. Ideally, we're just toggling the view of the summary or main content area inside the step.

**Review by nbloomf — APPROVED** [review]
Works as described. Nice!

File: packages/wp-checkout/src/components/checkout.js (hunk 4 above)
Comment (sirbrillig) [reply packages/wp-checkout/src/components/checkout.js]: Agreed. This was the result of a poor design decision on my part: I had originally forgotten that we need to have both displayed at the same time.

Comment (sirbrillig) [reply packages/wp-checkout/src/components/checkout.js]: Fixed by #36896

**Comment by simison** [comment]
Would it work to have package named just `@automattic/checkout`? Simpler. :-)

File: packages/composite-checkout/.prettierrc [hunk 5]
```diff
@@ -0,0 +1,10 @@
+useTabs: true
+tabWidth: 2
+printWidth: 100
+singleQuote: true
+trailingComma: es5
+bracketSpacing: true
+parenSpacing: true
+jsxBracketSameLine: false
+semi: true
+arrowParens: avoid
```
Comment (simison) [inline packages/composite-checkout/.prettierrc]: Why separate Prettier config from Calypso's own that's already there?

File: packages/composite-checkout/README.md [hunk 6]
```diff
@@ -0,0 +1,559 @@
+# WP Checkout
```
Comment (simison) [inline packages/composite-checkout/README.md]: This title could be same as package name - WP prefix feels confusing here.

File: packages/composite-checkout/README.md [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+
+A set of React components, custom Hooks, and helper functions that together can be used to create a purchase and checkout flow.
+
+## 💰 Installation
```
Comment (simison) [inline packages/composite-checkout/README.md]: I'd keep emojis out from titles, but I'm just little boring... :-D

In some systems these titles will form automatic # links. With emojis those will look funky.

File: packages/composite-checkout/README.md [hunk 8]
```diff
@@ -0,0 +1,559 @@
...
+
+### Submitting the form
+
+When the payment button is pressed, the form data will be validated and submitted in a way appropriate to the payment method. If there is a problem with either validation or submission, or if the payment method's service returns an error, the `onFailure` prop on `Checkout` will be called with an object describing the error.
+
+If the payment method succeeds, the `onSuccess` prop will be called instead. It's important not to provision anything based on this callback, as it could be called by a malicious user. Instead, provisioning should be handled separately server-to-server.
+
+Some payment methods may require a redirect to an external site. If that occurs, the `failureRedirectUrl` and `successRedirectUrl` props on `Checkout` will be used instead of the `onFailure` and `onSuccess` callbacks. All four props are required.
+
+## 💰 Example
```
Comment (simison) [inline packages/composite-checkout/README.md]: "Examples"?

File: packages/composite-checkout/README.md [hunk 9]
```diff
@@ -0,0 +1,559 @@
...
+
+### Can I add custom properties to line items?
+
+To maintain the integrity of the line item schema, adding custom fields is discouraged, but allowed. If you need specific custom data as part of a line item so that it can be used in another part of the form, it's recommended to pass the line item object through a helper function to collect the extra data rather than serializing it into the line item itself. This will make that data collection more testable. However, if the data collection process is expensive or slow, and caching isn't an option, it may make sense to preload the data into the line items.
+
+## 💰 Development
+
+In the root of the monorepo, run `npm run composite-checkout-demo` which will start a local webserver that will display the component.
```
Comment (simison) [inline packages/composite-checkout/README.md]: This is fine — incorporating to devdocs where rest of the stuff in calypso is showcased/developed would feel more aligned but not a Biggie. Prolly faster this way?

File: packages/composite-checkout/.gitignore [hunk 10]
```diff
@@ -0,0 +1 @@
+dist
```
Comment (simison) [inline packages/composite-checkout/.gitignore]: This should be gitignored already in the monorepo.

File: webpack.config.composite-config-demo.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
+const path = require( 'path' );
+const webpack = require( 'webpack' );
+
+module.exports = {
```
Comment (simison) [inline webpack.config.composite-config-demo.js]: You can use calypso build package to build so no need custom webpack and Babel configs here. See other /apps or /packages for examples or docs at calypso-build package.

File: packages/composite-checkout/src/theme.js [hunk 12]
```diff
@@ -0,0 +1,37 @@
+/* @format */
```
Comment (simison) [inline packages/composite-checkout/src/theme.js]: Our prettier setup now formats without these format pragmas so they can be removed.

File: packages/composite-checkout/src/lib/payment-methods/apple-pay.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+	const { billingName = '', billingNameError = null } = paymentData || {};
+	const onChange = value => setPaymentData( { ...( paymentData || {} ), billingName: value } );
+	return (
+		<BillingFormFields>
+			<FormField
+				id="billingName"
+				type="Text"
+				label="Name"
+				error={ billingNameError }
+				errorMessage="This is a required field"
```
Comment (simison) [inline packages/composite-checkout/src/lib/payment-methods/apple-pay.js]: Prolly needs i18n wrapper?

File: packages/composite-checkout/src/lib/localize.js [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+ * External dependencies
+ */
+import { useContext } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import CheckoutContext from './checkout-context';
+
+// TODO: we need to implement this; probably delegated to i18n-calypso
```
Comment (simison) [inline packages/composite-checkout/src/lib/localize.js]: Would using wp-i18n work instead? That can be externalised for core since it ships in WordPress so saves the bundle size in wp-admin.

**Comment by simison** [comment]
How using styled components package works with our aim to use `@wordpress/components` all across the board? Will they look out of place compared to rest of UI? Will that need significant rewrite later on?

File: webpack.config.composite-config-demo.js (hunk 11 above)
Comment (sirbrillig) [reply webpack.config.composite-config-demo.js]: This file is just to run the demo. I don't think we have anything like this so far, and because the demo script is in the global package.json I thought maybe this should be here. But perhaps it should be a second webpack config inside the package?

I considered putting this demo in devdocs, but since this has no calypso dependencies, it seems like a poor experience to have to load and rebuild all of calypso just to develop this package. I asked in Slack and it was suggested that I create a custom script for the demo. The better solution would be a devdocs that does not depend on calypso. I started down that road but then was told about #33166 which I think is the better solution. However, we can't wait to work on this project until Storybook is available, so a custom dev web server seems like the best temporary solution.

Comment (sirbrillig) [reply webpack.config.composite-config-demo.js]: I also may not be understanding your comment correctly 😓 ... reading the `calypso-build` README doesn't fully clarify to me how I would use it to run this dev server.

File: packages/composite-checkout/README.md (hunk 6 above)
Comment (sirbrillig) [reply packages/composite-checkout/README.md]: Oops. That's left over from the original package name. Thanks!

File: packages/composite-checkout/src/lib/localize.js (hunk 14 above)
Comment (sirbrillig) [reply packages/composite-checkout/src/lib/localize.js]: That's a good idea. I'll take a look when we get to this part.

**Comment by sirbrillig** [comment]
> How using styled components package works with our aim to use @wordpress/components all across the board? Will they look out of place compared to rest of UI? Will that need significant rewrite later on?

This is a good question. We discussed this (p4jECJ-1of-p2) and the answer seemed to be that styled-components may be the future of CSS for gutenberg (https://github.com/WordPress/gutenberg/pull/17614) and a8c projects (pbjpUB-1O-p2, pb3aDo-85-p2). It also provides a very elegant way to allow theming. That's why we went with this approach. However, other opinions are welcome!

File: packages/composite-checkout/README.md (hunk 9 above)
Comment (sirbrillig) [reply packages/composite-checkout/README.md]: Yes. I did consider this, but devdocs is quite slow and loads a lot of unnecessary things (ie: calypso) for this project. The best solution would be Storybook (#33166) but this is our temporary solution until we have that. We could also put it in devdocs once we are not in active development.

**Comment by sirbrillig** [comment]
Merging this now since there have been no new comments for several days and other PRs are piling up behind this. I think it's safe; if we want to make any further changes, they can be done in additional PRs.
